diff-test/core/src/model.ts

```diff
@@ -0,0 +1,734 @@
+   * A packet is deliverable if it two blocks have committed on
+   * the sender chain since the packet was sent. This is as
+   * per the light-client functioning.
+   * 
+   * @returns The number of deliverable packets.
+   */
+  numAvailable = (): number => {
+    return this.fifo.filter((e) => 1 < e[1]).length;
+  };
+  /**
+   * Get and internally delete deliverable packets from the outbox.
+   * 
+   * @param num num packets to consumer
+   * @returns A list of deliverable packets
+   */
+  consume = (num: number): Packet[] => {
+    const [available, unavailable] = _.partition(
+      this.fifo,
+      (e) => 1 < e[1],
+    );
+    const take = available.slice(0, num);
+    this.fifo = available.slice(num).concat(unavailable);
+    return take.map((e) => e[0]);
+  };
+  /**
+   * Commit the packets in the outbox. Once a packet has been
+   * committed twice it is available for delivery, as per the
+   * ibc light-client functioning.
+   */
+  commit = () => {
+    this.fifo = this.fifo.map((e) => [e[0], e[1] + 1]);
+  };
+}
+
+class Staking {
+  // Model handle
+  m;
+  // Validator delegations from the delegator
+  // A fixed descending order is used for the initial values to allow
+  // easy setup in the SUT.
+  delegation = [
+    4 * TOKEN_SCALAR,
+    3 * TOKEN_SCALAR,
+    2 * TOKEN_SCALAR,
+    1 * TOKEN_SCALAR,
+  ];
+  // Validator tokens
+  // 1 additional unit is given to prevent validators from being deleted
+  // by the staking module when the delegation falls to 0.
+  tokens: number[] = this.delegation.map((it) => it + 1 * TOKEN_SCALAR);
+  // Validator status
+  status = [
+    Status.BONDED, // Bonded as per the delegation above
+    Status.BONDED, // ^^
+    Status.UNBONDED, // Unbonded as per MAX_VALIDATORS
+    Status.UNBONDED, // ^^
+  ];
+  // Undelegation queue
+  undelegationQ: Undelegation[] = [];
+  // Unbonding validator queue
+  validatorQ: Unval[] = [];
+  // Validator jail timestamp
+  // Undefined if validator is not jailed
+  jailed: number | undefined[] = new Array(NUM_VALIDATORS).fill(
+    undefined,
+  );
+  // Initial balance of delegator account.
+  delegatorTokens: number = INITIAL_DELEGATOR_TOKENS;
+  // Unique ID used to count unbonding and redelegation queue entries,
+  // as well as unbonding validators.
+  opID = 0;
+  // maps validator id -> power
+  // used to compute validator set changes
+  changes = {};
+  // The validators of the last block
+  lastVals;
+  // The number of tokens of the last block
+  // Used to compute validator power changes used in VSCs
+  lastTokens = _.clone(this.tokens);
+
+  /**
+   * Compute the new set of active validators
+   */
+  newVals = () => {
+    const valid = (i): boolean =>
+      1 <= this.tokens[i] && this.jailed[i] === undefined;
+    let vals = _.range(NUM_VALIDATORS);
+    // stable sort => breaks ties based on validator
+    // address numerical value. This mimics staking module.
+    vals = _.sortBy(vals, (i) => -this.tokens[i]);
```

Comment:
Why not use the built in `.sort`?

Reply:
That doesn't sort numbers

Reply:
Yep, good point. It's probably like this because I was using sort at one point, not realising it did string conversion and then when I fixed it didn't check if there was an optional lambda arg.

Reply:
Actually, no, just remembered. I did it because underscore sort is guaranteed stable. I didn't see stable guarantee written anywhere in js version.